addons/website/models/website_visitor.py

```diff
@@ -0,0 +1,140 @@
+
+    def _get_visitor_from_request(self):
+        if not request:
+            return None
+        visitor = self.env['website.visitor']
+        cookie_content = request.httprequest.cookies.get('visitor_id')
+        if cookie_content and '-' in cookie_content:
+            visitor_id, visitor_hash = cookie_content.split('-', 1)
+            if _consteq(visitor_hash, self._get_visitor_hash(visitor_id)):
+                return visitor.sudo().with_context(active_test=False).search([('id', '=', visitor_id)])  # search to avoid having to call exists()
+        return visitor
+
+    def _handle_webpage_dispatch(self, response, website_page):
+        if website_page:
+            # get visitor only if page tracked. Done here to avoid having to do it multiple times in case of override.
+            visitor_sudo = self._get_visitor_from_request() if website_page.is_tracked else False
+            self._handle_website_page_visit(response, website_page, visitor_sudo)
+
+    def _handle_website_page_visit(self, response, website_page, visitor_sudo):
+        """ Called on dispatch. This will create a website.visitor if the http request object
+        is a tracked website page. Only on tracked page to avoid having too much operations done on every page
+        or other http requests.
+        Note: The side effect is that the last_connection_datetime is updated ONLY on tracked pages."""
+        if website_page.is_tracked:
+            if not visitor_sudo:
+                # If visitor does not exist
+                visitor_sudo = self._create_visitor(website_page.id)
+                sign = visitor_sudo._get_visitor_sign().get(visitor_sudo.id)
+                response.set_cookie('visitor_id', sign)
+            else:
+                # Add page even if already in visitor_page_ids as checks on relations are done in many2many write method
+                vals = {
+                    'last_connection_datetime': datetime.now(),
+                    'visitor_page_ids': [(0, 0, {'page_id': website_page.id, 'visit_datetime': datetime.now()})],
+                }
+                if visitor_sudo.last_connection_datetime < (datetime.now() - timedelta(hours=8)):
+                    vals['visit_count'] = visitor_sudo.visit_count + 1
+                if not visitor_sudo.active:
+                    vals['active'] = True
+                visitor_sudo.write(vals)
+
+    def _create_visitor(self, website_page_id=False):
+        country_code = request.session.get('geoip', {}).get('country_code', False)
+        country_id = request.env['res.country'].sudo().search([('code', '=', country_code)], limit=1).id if country_code else False
+        lang_id = request.env['res.lang'].sudo().search([('code', '=', request.lang)], limit=1).id
+        vals = {
+            'last_connection_datetime': datetime.now(),
+            'lang_id': lang_id,
+            'country_id': country_id,
+            'website_id': request.website.id
+        }
```

Comment:
If a connected user go to a tracked page for the first time he is not linked to the visitor. `user_partner_ids` should be set if the user is not public.

Reply:
you're totally right ! Thanks !